release_util/management/commands/__init__.py

```diff
@@ -132,164 +134,215 @@ def state(self):
-                        self.migration_state['failure'] = {
-                            'migration': migration,
-                            'output': out.getvalue(),
-                            'traceback': repr(traceback.format_exception(exc_type, exc_value, exc_traceback)),
-                        }
-                        # Save the total migration time until failure.
-                        self.migration_state['duration'] = time_to_fail
-                        break
-
-            # Remove any migrations that succeeded -or- failed.
-            # Any leftover migrations were *not* applied.
-            succeeded = [status['migration'] for status in self.migration_state['success']]
-            failed = []
-            if failed_app and failed_migration:
-                failed = [failed_app, failed_migration]
-            for migration in self.to_apply:
-                if migration not in succeeded and migration != failed:
-                    self.migration_state['unapplied'].append(migration)
+        "Unapplied" is a list of unapplied migrations. "Current" is a list of the current migration
+        states for apps with unapplied migrations.
 
-            # Find the apps that have been applied -or- failed.
-            # Include their initial migrations as commands.
-            self._add_rollback_commands()
-            raise CommandError("Migration failed for app '{}' - migration '{}'.\n".format(failed_app, failed_migration))
+        Both are tuples of the form (app, migration_name).
+        """
+        connection = connections[self._database_name]
+        loader = MigrationLoader(connection, ignore_no_migrations=True)
+        unapplied = self._gather_unapplied_migrations(loader)
+        currents = self._gather_current_migration_state(loader, [u[0] for u in unapplied])
+        return unapplied, currents
 
-        # All migrations succeeded. Record the total time for all migrations.
-        time_to_apply = self.timer() - start
-        self.migration_state['duration'] = time_to_apply
+    def _parse_migrate_output(self, output):
+        """
+        Returns (succeeded:list(tuple), failed:tuple or None)
+        Both tuples are of the form (app, migration)
+        """
+        failed = None
+        succeeded = []
 
-        # Parse the output to find all migrations that were applied successfully.
-        for line in out.getvalue().split('\n'):
+        # Mark migrations:
+        # - before exception migration as success
+        # - exception migration as failed
+        # - any remaining exceptions as unapplied
+        for line in output.split('\n'):
             line = _remove_escape_characters(line).strip()
             line_match = self.migration_regex.match(line)
             if line_match:
-                migration = [line_match.group('app_name'), line_match.group('migration_name')]
+                migration = (line_match.group('app_name'), line_match.group('migration_name'))
```

Comment:
I was thinking about migrations as sort of an ad-hoc internal datatype along the lines of `namedtuple('Migration', ['app', 'migration'])`. (If I had more time I'd go and actually use namedtuples). That sense was already present in the code, but I tried to standardize on it where I saw it used. I chose tuples over lists because this data feels more tuple-y (fixed-length, immutable, non-homogenous, etc.)

What are your concerns relating to serialization of namedtuples? They'll just get output as YAML lists. If you'd prefer the explicitness of YAML dicts in the output, we could do that in the management command itself, or make an actual Migration class that serializes to a dict.